File: packages/domain/src/math.ts
import { fail } from "./errors.ts";

/** Fixed-width integer ranges of the onchain program. Amounts are `bigint`, never `number`. */
export const U8_MAX = 255n;
export const U16_MAX = 65_535n;
export const U32_MAX = 4_294_967_295n;
export const U64_MAX = 18_446_744_073_709_551_615n;
export const I64_MAX = 9_223_372_036_854_775_807n;
export const I64_MIN = -9_223_372_036_854_775_808n;

export const BPS_DENOM = 10_000n;

export const isU32 = (x: bigint): boolean => x >= 0n && x <= U32_MAX;
export const isU64 = (x: bigint): boolean => x >= 0n && x <= U64_MAX;
export const isI64 = (x: bigint): boolean => x >= I64_MIN && x <= I64_MAX;

/** Narrow to u64 or fail with `ArithmeticOverflow`. */
export function asU64(x: bigint): bigint {
  return isU64(x) ? x : fail("ArithmeticOverflow", "value outside u64");
}

/** Narrow to i64 or fail with `ArithmeticOverflow`. */
export function asI64(x: bigint): bigint {
  return isI64(x) ? x : fail("ArithmeticOverflow", "value outside i64");
}

export const addU64 = (a: bigint, b: bigint): bigint => asU64(asU64(a) + asU64(b));
export const subU64 = (a: bigint, b: bigint): bigint => asU64(asU64(a) - asU64(b));
export const mulU64 = (a: bigint, b: bigint): bigint => asU64(asU64(a) * asU64(b));

export function divU64(a: bigint, b: bigint): bigint {
  if (asU64(b) === 0n) return fail("DivisionByZero");
  return asU64(a) / b;
}

export function ceilDivU64(a: bigint, b: bigint): bigint {
  if (asU64(b) === 0n) return fail("DivisionByZero");
  asU64(a);
  return (a + b - 1n) / b;
}

/**
 * floor(a * b / c) with a widened (unbounded) intermediate, result must fit u64.
 * The widened product mirrors Rust's u128; a u64 x u64 product always fits u128.
 */
export function mulDivFloorU64(a: bigint, b: bigint, c: bigint): bigint {
  if (asU64(c) === 0n) return fail("DivisionByZero");
  return asU64((asU64(a) * asU64(b)) / c);
}

/** ceil(a * b / c), widened, result must fit u64. */
export function mulDivCeilU64(a: bigint, b: bigint, c: bigint): bigint {
  if (asU64(c) === 0n) return fail("DivisionByZero");
  return asU64((asU64(a) * asU64(b) + c - 1n) / c);
}

/** Parse a base-10 unsigned integer string (JSON transport form) into a bigint. */
export function parseUint(text: string): bigint {
  if (!/^(0|[1-9][0-9]*)$/.test(text))
    throw new TypeError("expected a base-10 unsigned integer string");
  return BigInt(text);
}

/** Parse a base-10 signed integer string into a bigint. */
export function parseInt64(text: string): bigint {
  if (!/^-?(0|[1-9][0-9]*)$/.test(text)) throw new TypeError("expected a base-10 integer string");
  return BigInt(text);
}

File: crates/mandate-core/src/math.rs
//! Checked fixed-width integer helpers. Nothing wraps and nothing saturates silently.
use crate::{MandateCoreError, Result};

pub fn add_u64(a: u64, b: u64) -> Result<u64> {
    a.checked_add(b).ok_or(MandateCoreError::ArithmeticOverflow)
}

pub fn sub_u64(a: u64, b: u64) -> Result<u64> {
    a.checked_sub(b).ok_or(MandateCoreError::ArithmeticOverflow)
}

pub fn mul_u64(a: u64, b: u64) -> Result<u64> {
    a.checked_mul(b).ok_or(MandateCoreError::ArithmeticOverflow)
}

pub fn div_u64(a: u64, b: u64) -> Result<u64> {
    a.checked_div(b).ok_or(MandateCoreError::DivisionByZero)
}

pub fn rem_u64(a: u64, b: u64) -> Result<u64> {
    a.checked_rem(b).ok_or(MandateCoreError::DivisionByZero)
}

/// `ceil(a / b)`.
pub fn ceil_div_u64(a: u64, b: u64) -> Result<u64> {
    let quotient = div_u64(a, b)?;
    if rem_u64(a, b)? == 0 {
        Ok(quotient)
    } else {
        // quotient < a / b + 1 <= u64::MAX + 1 only when b == 1, which has remainder 0, so this cannot overflow;
        // it is still checked.
        add_u64(quotient, 1)
    }
}

fn narrow_u64(value: u128) -> Result<u64> {
    u64::try_from(value).map_err(|_| MandateCoreError::ArithmeticOverflow)
}

/// `floor(a * b / c)` with a `u128` intermediate; the result must fit `u64`.
pub fn mul_div_floor_u64(a: u64, b: u64, c: u64) -> Result<u64> {
    let product = u128::from(a)
        .checked_mul(u128::from(b))
        .ok_or(MandateCoreError::ArithmeticOverflow)?;
    let quotient = product
        .checked_div(u128::from(c))
        .ok_or(MandateCoreError::DivisionByZero)?;
    narrow_u64(quotient)
}

/// `ceil(a * b / c)` with a `u128` intermediate; the result must fit `u64`.
pub fn mul_div_ceil_u64(a: u64, b: u64, c: u64) -> Result<u64> {
    let product = u128::from(a)
        .checked_mul(u128::from(b))
        .ok_or(MandateCoreError::ArithmeticOverflow)?;
    let divisor = u128::from(c);
    let quotient = product
        .checked_div(divisor)
        .ok_or(MandateCoreError::DivisionByZero)?;
    let remainder = product
        .checked_rem(divisor)
        .ok_or(MandateCoreError::DivisionByZero)?;
    let rounded = if remainder == 0 {
        quotient
    } else {
        quotient
            .checked_add(1)
            .ok_or(MandateCoreError::ArithmeticOverflow)?
    };
    narrow_u64(rounded)
}

/// Narrow an `i128` timestamp back to `i64`.
pub(crate) fn narrow_i64(value: i128) -> Result<i64> {
    i64::try_from(value).map_err(|_| MandateCoreError::ArithmeticOverflow)
}
